### ISSUE TITLE
Fail to pass the request data throw Java client

### ISSUE THREAD

**Issue by tobegit3hub** [grpc/grpc#8062]
Now we have implement the gRPC server and clients. The Python server and Python client work well but the Java client can not pass the request data.

The protobuf looks like this and we have to set the map data.

```
syntax = "proto3";

package tensorflow.serving;
option cc_enable_arenas = true;

import "tensorflow/core/framework/tensor.proto";

message PredictRequest {
  map<string, TensorProto> inputs = 1;
}

message PredictResponse {
  map<string, TensorProto> outputs = 1;
}

service PredictionService {
  rpc Predict(PredictRequest) returns (PredictResponse);
}
```

And the client code of Java looks like this.

```
    tensorflow.serving.Predict.PredictRequest.Builder request_builder = tensorflow.serving.Predict.PredictRequest.newBuilder().putInputs("features", tensorProto);
    Predict.PredictRequest request = request_builder.build();

    if(request_builder.containsInputs("features")) {
      System.out.print("Container the features");
    } else {
      System.out.print("Don'et contain the features");
    }
```

I'm pretty sure that it prints out "Container the features". But when I trace the `request` in server side, the `request.inputs` is `{}`. If I use the Python client, it works.

```
    def Predict(self, request, context):
        import ipdb;ipdb.set_trace()
        request_map = request.inputs
```

I'm not sure if it's the bug of Java implementation. It would be great if anyone has any example to work with Java.

**Comment by tobegit3hub** [comment]
It seems like a bug of gRPC. The source code is in https://github.com/tobegit3hub/deep_recommend_system/tree/master/grpc_service/java and I hope it helps to fix it ASAP.

**Comment by tobegit3hub** [comment]
Now I try to implement the server side in Java. The code looks like this and it got two `inputs` from request now.

```
  private class PredictionServiceImpl extends PredictionServiceGrpc.PredictionServiceImplBase {

    @Override
    public void predict(Predict.PredictRequest req, StreamObserver<Predict.PredictResponse> responseObserver) {
      System.out.print("Inputs count: " + Integer.toString(req.getInputsCount()));

      Predict.PredictResponse reply = Predict.PredictResponse.newBuilder().build();
      responseObserver.onNext(reply);
      responseObserver.onCompleted();
    }
  }
```

The output is 2 when I request from Java client. But If I request with my previous Python client, it printed 0 now. That means the Python client can work with Python server and Java client can work with Java server, but not both :(

**Comment by vjpai** [comment]
@nathanielmanistaatgoogle @ejona86 : please comment if this should still be open.

**Comment by nathanielmanistaatgoogle** [comment]
There's not enough here for me to go on; I'd like to hear @ejona86's assessment.

@tobegit3hub: is this still an unsolved problem for you? We have an interoperation test suite that passes every day; we strongly believe that a Python server and Java client can interact perfectly well.

**Comment by tobegit3hub** [comment]
Sorry about I don't have the environment to test again. It's okay to close if you have well tested.

**Comment by ejona86** [comment]
If anything, I would suspect protobuf. Our interop tests don't use `map`, so if it was incompatible between Python and Java we probably wouldn't notice. We would likely notice grpc-caused corruption in the interop tests.

**Comment by nathanielmanistaatgoogle** [comment]
@haberman, how confident do we feel about the compatibility of `map` serialization and deserialization between Java and Python?

**Comment by haberman** [comment]
Map serialization should be very consistent and interoperable. If you can reproduce a bug in serialization between languages, please file an issue against protobuf. Ideally the test case would not involve gRPC but would just be straight protobuf.

**Comment by nathanielmanistaatgoogle** [comment]
@haberman: I agree with "should be"; does protobuf have test coverage verifying this particular case?

**Comment by nathanielmanistaatgoogle** [comment]
This seems to be low-priority; punting. We can always raise the priority later if someone steps forward as being affected by a defect here.

**Comment by anshulkjain** [comment]
This is not only the case with map, I'm facing this with arrays as well. First I tried with map and when it didn't work modified my proto to use arrays but same issue occurred. As it's said in this email chain before that python clients works with python server and java client with java server but interoperability between both the languages failed while using map / array. Can someone advise if it's gonna be fixed anytime soon?
